File: Assets/Bundles/Shared/Resources/TS/Effects/EffectsManager.lua
-- Compiled with unity-ts v2.1.0-75
local GameObjectBridge = require("Shared/TS/GameObjectBridge").GameObjectBridge
local BundleReferenceManager = require("Shared/TS/Util/BundleReferenceManager").BundleReferenceManager
local ReferenceManagerAssets = require("Shared/TS/Util/ReferenceManagerResources").ReferenceManagerAssets
local EffectsManager
do
	EffectsManager = setmetatable({}, {
		__tostring = function()
			return "EffectsManager"
		end,
	})
	EffectsManager.__index = EffectsManager
	function EffectsManager.new(...)
		local self = setmetatable({}, EffectsManager)
		return self:constructor(...) or self
	end
	function EffectsManager:constructor()
	end
	function EffectsManager:SpawnBundleEffect(bundleGroupId, bundleId, effectId, worldPosition, worldRotation, destroyInSeconds)
		if destroyInSeconds == nil then
			destroyInSeconds = 5
		end
		local _bundleGroups = ReferenceManagerAssets.bundleGroups
		local _bundleGroupId = bundleGroupId
		local bundleGroup = _bundleGroups[_bundleGroupId]
		if bundleGroup then
			return self:SpawnBundleGroupEffect(bundleGroup, bundleId, effectId, worldPosition, worldRotation, destroyInSeconds)
		end
		return nil
	end
	function EffectsManager:SpawnBundleGroupEffect(bundleGroup, bundleId, effectId, worldPosition, worldRotation, destroyInSeconds)
		if destroyInSeconds == nil then
			destroyInSeconds = 5
		end
		local _bundles = bundleGroup.bundles
		local _bundleId = bundleId
		local bundle = _bundles[_bundleId]
		if not bundle then
			return nil
		end
		local effect = self:SpawnBundleDataEffect(bundle, effectId, nil, destroyInSeconds)
		if effect then
			effect.transform.position = worldPosition
			effect.transform.eulerAngles = worldRotation
		end
		return effect
	end
	function EffectsManager:SpawnBundleDataEffect(bundle, effectId, hitTransform, destroyInSeconds)
		if destroyInSeconds == nil then
			destroyInSeconds = 5
		end
		if not bundle or effectId < 0 then
			error("Trying to spawn effect that doesnt exist: " .. tostring(bundle) .. ", " .. tostring(effectId))
			return nil
		end
		local template = BundleReferenceManager:LoadResourceFromBundle(bundle, effectId)
		if template == nil then
			error("Trying to spawn effect but prefab template wasn't found: " .. tostring(bundle.id) .. ", " .. tostring(effectId))
			return nil
		end
		return self:SpawnEffect(template, hitTransform, destroyInSeconds)
	end
	function EffectsManager:SpawnEffectAtPosition(template, worldPosition, worldEuler, destroyInSeconds)
		if destroyInSeconds == nil then
			destroyInSeconds = 5
		end
		local effect = self:SpawnEffect(template, nil, destroyInSeconds)
		effect.transform.position = worldPosition
		if worldEuler then
			effect.transform.eulerAngles = worldEuler
		end
		return effect
	end
	function EffectsManager:SpawnEffect(template, parent, destroyInSeconds)
		if destroyInSeconds == nil then
			destroyInSeconds = 5
		end
		local vfx
		if parent then
			vfx = GameObjectBridge:InstantiateIn(template, parent)
		else
			vfx = GameObjectBridge:Instantiate(template)
		end
		vfx.transform.localPosition = Vector3.zero
		vfx.transform.localEulerAngles = Vector3.zero
		-- vfx.transform.localScale = Vector3.one;
		if destroyInSeconds > 0 then
			GameObjectBridge:Destroy(vfx, destroyInSeconds)
		end
		return vfx
	end
end
return {
	EffectsManager = EffectsManager,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Shared/Effects/EffectsManager.ts
import { WorldAPI } from "Shared/VoxelWorld/WorldAPI";
import { BundleReferenceManager } from "../Util/BundleReferenceManager";
import {
	AllBundleItems,
	BundleData,
	BundleGroup,
	BundleGroupNames,
	ReferenceManagerAssets,
} from "../Util/ReferenceManagerResources";
import { Task } from "Shared/Util/Task";
import { GameObjectUtil } from "Shared/GameObject/GameObjectUtil";

export class EffectsManager {
	public static SpawnEffect(
		bundleId: AllBundleItems,
		worldPosition: Vector3,
		worldRotation: Vector3,
		destroyInSeconds = 5,
	) {
		let template = BundleReferenceManager.LoadDirectResource<GameObject>(bundleId);
		if (template) {
			return this.SpawnGameObjectAtPosition(template, worldPosition, worldRotation, destroyInSeconds);
		}
		return undefined;
	}

	public static SpawnBundleEffect(
		bundleGroupId: BundleGroupNames,
		bundleId: number,
		effectId: number,
		worldPosition: Vector3,
		worldRotation: Vector3,
		destroyInSeconds = 5,
	) {
		let bundleGroup = ReferenceManagerAssets.bundleGroups.get(bundleGroupId);
		if (bundleGroup) {
			return this.SpawnBundleGroupEffect(
				bundleGroup,
				bundleId,
				effectId,
				worldPosition,
				worldRotation,
				destroyInSeconds,
			);
		}
		return undefined;
	}

	public static SpawnBundleGroupEffect(
		bundleGroup: BundleGroup,
		bundleId: number,
		effectId: number,
		worldPosition: Vector3,
		worldRotation: Vector3,
		destroyInSeconds = 5,
	) {
		let bundle = bundleGroup.bundles.get(bundleId);
		if (!bundle) {
			return undefined;
		}
		let effect = this.SpawnBundleDataEffect(bundle, effectId, undefined, destroyInSeconds);
		if (effect) {
			effect.transform.position = worldPosition;
			effect.transform.eulerAngles = worldRotation;
		}

		return effect;
	}

	public static SpawnBundleDataEffect(
		bundle: BundleData,
		effectId: number,
		hitTransform: Transform | undefined,
		destroyInSeconds = 5,
	): GameObject | undefined {
		if (!bundle || effectId < 0) {
			error("Trying to spawn effect that doesnt exist: " + bundle + ", " + effectId);
			return;
		}
		let template = BundleReferenceManager.LoadResourceFromBundle<GameObject>(bundle, effectId);
		if (template === undefined) {
			error("Trying to spawn effect but prefab template wasn't found: " + bundle.id + ", " + effectId);
			return undefined;
		}
		return this.SpawnGameObject(template, hitTransform, destroyInSeconds);
	}

	public static SpawnGameObjectAtPosition(
		template: GameObject,
		worldPosition: Vector3,
		worldEuler?: Vector3,
		destroyInSeconds = 5,
	) {
		let effect = this.SpawnGameObject(template, undefined, destroyInSeconds);
		effect.transform.position = worldPosition;
		if (worldEuler) {
			effect.transform.eulerAngles = worldEuler;
		}
		return effect;
	}

	public static SpawnGameObject(template: GameObject, parent?: Transform, destroyInSeconds = 5) {
		let vfx: GameObject;
		vfx = PoolManager.SpawnObject(template);
		if (parent) {
			vfx.transform.SetParent(parent);
		}
		vfx.transform.localPosition = Vector3.zero;
		vfx.transform.localEulerAngles = Vector3.zero;
		//vfx.transform.localScale = Vector3.one;

		if (destroyInSeconds > 0) {
			Task.Delay(destroyInSeconds, () => {
				PoolManager.ReleaseObject(vfx);
			});
		}
		return vfx;
	}

	public static ReleaseGameObject(go: GameObject) {
		PoolManager.ReleaseObject(go);
	}

	public static SetParticleToBlockMaterial(particles: ParticleSystemRenderer, blockId: number) {
		const world = WorldAPI.GetMainWorld();
		if (!world) return;
		const blockGO = MeshProcessor.ProduceSingleBlock(blockId, world.voxelWorld);
		if (blockGO) {
			const blockRen = blockGO.GetComponent<Renderer>();
			const blockFilter = blockGO.GetComponent<MeshFilter>();
			particles.mesh = blockFilter.mesh;
			particles.sharedMaterial = blockRen.sharedMaterial;
			GameObjectUtil.Destroy(blockGO);
		}
	}
}
